commit cb7ab7854bc70942abed62d19d8c16d0064bf7dc
Author: Justin Stitt <justinstitt@google.com>
Date:   Thu Sep 21 07:58:48 2023 +0000

    IB/qib: Replace deprecated strncpy
    
    `strncpy` is deprecated for use on NUL-terminated destination strings [1]
    and as such we should prefer more robust and less ambiguous string
    interfaces.
    
    We know `txselect_list` is expected to be NUL-terminated based on its
    use in `param_get_string()`:
    | int param_get_string(char *buffer, const struct kernel_param *kp)
    | {
    |       const struct kparam_string *kps = kp->str;
    |       return scnprintf(buffer, PAGE_SIZE, "%s\n", kps->string);
    | }
    
    Note that `txselect_list` is assigned to `kp_txselect`'s string field:
    | static struct kparam_string kp_txselect = {
    |       .string = txselect_list,
    |       .maxlen = MAX_ATTEN_LEN
    | };
    
    Wherein it is then assigned the set and get methods:
    | module_param_call(txselect, setup_txselect, param_get_string,
    |                 &kp_txselect, S_IWUSR | S_IRUGO);
    
    Considering the above, a suitable replacement is `strscpy` [2] due to
    the fact that it guarantees NUL-termination on the destination buffer
    without unnecessarily NUL-padding.
    
    Link: https://www.kernel.org/doc/html/latest/process/deprecated.html#strncpy-on-nul-terminated-strings [1]
    Link: https://manpages.debian.org/testing/linux-manual-4.8/strscpy.9.en.html [2]
    Link: https://github.com/KSPP/linux/issues/90
    Cc: linux-hardening@vger.kernel.org
    Signed-off-by: Justin Stitt <justinstitt@google.com>
    Link: https://lore.kernel.org/r/20230921-strncpy-drivers-infiniband-hw-qib-qib_iba7322-c-v1-1-373727763f5b@google.com
    Signed-off-by: Leon Romanovsky <leon@kernel.org>

diff --git a/drivers/infiniband/hw/qib/qib_iba7322.c b/drivers/infiniband/hw/qib/qib_iba7322.c
index 9d2dd135b784..f93906d8fc09 100644
--- a/drivers/infiniband/hw/qib/qib_iba7322.c
+++ b/drivers/infiniband/hw/qib/qib_iba7322.c
@@ -6127,7 +6127,7 @@ static int setup_txselect(const char *str, const struct kernel_param *kp)
 			TXDDS_TABLE_SZ + TXDDS_EXTRA_SZ + TXDDS_MFG_SZ);
 		return -EINVAL;
 	}
-	strncpy(txselect_list, str, ARRAY_SIZE(txselect_list) - 1);
+	strscpy(txselect_list, str, sizeof(txselect_list));
 
 	xa_for_each(&qib_dev_table, index, dd)
 		if (dd->deviceid == PCI_DEVICE_ID_QLOGIC_IB_7322)